chat/src/client.rs

```diff
@@ -0,0 +1,93 @@
+use crate::protocol::{ChatMessage, MemberRequest, MemberResponse};
+use crate::Opts;
+use hydroflow::builder::prelude::*;
+use hydroflow::scheduled::graph_ext::GraphExt;
+use hydroflow::scheduled::handoff::VecHandoff;
+
+pub(crate) async fn run_client(opts: Opts) {
+    let mut df = HydroflowBuilder::default();
+
+    // setup connection req/resp ports
+    let connect_req = df.hydroflow.outbound_tcp_vertex::<MemberRequest>().await;
+    let connect_req = df.wrap_output(connect_req);
+
+    let (connect_response_port, connect_resp) =
+        df.hydroflow.inbound_tcp_vertex::<MemberResponse>().await;
+    let connect_resp = df.wrap_input(connect_resp);
+
+    // setup message send/recv ports
+    let (messages_port, messages_recv) = df.hydroflow.inbound_tcp_vertex::<ChatMessage>().await;
+    let messages_recv = df.wrap_input(messages_recv);
+    let messages_send = df.hydroflow.outbound_tcp_vertex().await;
+    let messages_send = df.wrap_output(messages_send);
+
+    // setup stdio input handler
+    let (text_in, text_out) = df.add_channel_input::<Option<_>, VecHandoff<String>>();
+    hydroflow::tokio::spawn(async move {
+        // this boilerplate from tokio seems like a bummer to have in every program that needs kb input
+        // TODO(mingwei): temp, need to integrate stream into surface API.
+        use hydroflow::tokio::io::AsyncBufReadExt;
+
+        let mut reader = hydroflow::tokio::io::BufReader::new(hydroflow::tokio::io::stdin());
+        let mut buf = String::new();
+        while let Ok(_num_chars) = reader.read_line(&mut buf).await {
+            text_in.give(Some(buf.clone()));
+            text_in.flush();
+            buf.clear(); // TODO(mingwei): Maybe not needed?
+        }
+    });
```

Comment:
TODO(mingwei): once streams are integrated into the surface API this whole block can become a single method call.
```rust
let text_out = df.add_stream_input(hydroflow::tokio::io::stdin());
```